alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/healthprocessor/plugins/solr/SolrSearchExecutor.java

```diff
@@ -48,73 +44,97 @@ public SolrSearchExecutor() {
     public SolrSearchResult checkNodeIndexed(SearchEndpoint endpoint, Collection<Status> nodeStatuses)
             throws IOException {
 
-        String dbIdsQuery = nodeStatuses.stream()
-                .map(Status::getDbId)
-                .map(dbId -> "DBID:" + dbId)
-                .collect(Collectors.joining("%20OR%20"));
-
-        log.debug("Search query to endpoint {}: {}", endpoint, dbIdsQuery);
-
-        HttpUriRequest searchRequest = new HttpGet(
-                endpoint.getBaseUri()
-                        .resolve("select?q=" + dbIdsQuery + "&fl=DBID&wt=json&rows=" + nodeStatuses.size()));
-
-        log.trace("Executing HTTP request {}", searchRequest);
-        JsonNode response = httpClient.execute(searchRequest, new JSONResponseHandler());
+        // Initially, try a fetch for double the size of the node statuses array
+        // This is so we can immediately detect the case where all nodes are indexed twice.
+        int fetchSize = nodeStatuses.size()*2;
+        JsonNode response = executeSearchRequest(endpoint, nodeStatuses, fetchSize);
+
+        long numberOfFoundDocs = response.path("response").path("numFound").asLong();
+        if(numberOfFoundDocs > fetchSize) {
+            // We did not fetch enough in one batch to fetch all duplicates (when nodes are duplicated more than once)
+            // Send a new request for the number of rows we actually need.
+            log.debug("Found number of docs #{} is larger than the requested number of rows #{}. Fetching again with larger number of rows.", numberOfFoundDocs, fetchSize);
+            response = executeSearchRequest(endpoint ,nodeStatuses, numberOfFoundDocs);
```

Comment:
Hmm, good point.

I am not sure that this will cause that problem. We go to solr nodes directly, we do not follow the weird path that the search API follows with bouncing through one of the solr nodes to get all sharded results.

I think it is not to hard to split our nodes in batches once we pass a certain threshold.

Reply:
Hmm, implementing that batching proves to be more difficult than I expected, especially since we want to avoid sending requests to solr that we can avoid.

To implement batching properly need to do something like this:

1. Send a request to solr with the number of nodes that is the maximum batch size
2. If we have numDocs returned that is larger than what we expected:
     1. split into batches
     2. To know how big these batches may be, we need to guess how to split it up. Let's say we split up nodes equally.
     3. perform search again for each of those batches, go back to (1)
3. If we have numDocs returned that is equal or smaller than what we expected: process as normal.

it is quite complicated to do this right, and as you can see, we will send many requests to solr in some cases.

The default batch size of health-processor is 1000 nodes, so unless we have *a lot* of duplicate nodes, I don't expect a problem with this.
If we run into problems here, that means that the solr index is definitely messed up very badly.
The administrator can decrease the batch size to work around the problem immediately and still get a complete report.

If we ever run into this problem, we can pick this up and try our hand at batching. But as far as I am currently aware, we have a very small risk of ever running into this issue and we should not complicate the code further to handle it.

Reply:
> we should not complicate the code further to handle it.

Agreed. We should indeed normally not run into this.